ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/PartitionByParamsFactory.java

```diff
@@ -76,118 +79,168 @@ private PartitionByParamsFactory() {
+  // TODO: move, add accessors?
+  private static class PartitionByColumn {
+    final ColumnName name;
+    final boolean shouldAppend;
 
-      return Optional.of(column.name());
+    PartitionByColumn(final ColumnName name, final boolean shouldAppend) {
+      this.name = Objects.requireNonNull(name, "name");
+      this.shouldAppend = shouldAppend;
     }
+  }
 
-    return Optional.empty();
+  private static List<PartitionByColumn> getPartitionByColumnName(
+      final LogicalSchema sourceSchema,
+      final List<Expression> partitionByExpressions
+  ) {
+    final ColumnAliasGenerator columnAliasGenerator =
+        ColumnNames.columnAliasGenerator(Stream.of(sourceSchema));
+    return partitionByExpressions.stream()
+        .map(partitionBy -> {
+          if (partitionBy instanceof ColumnReferenceExp) {
+            // PARTITION BY column:
+            final ColumnName columnName = ((ColumnReferenceExp) partitionBy).getColumnName();
+
+            final Column column = sourceSchema
+                .findValueColumn(columnName)
+                .orElseThrow(() ->
+                    new IllegalStateException("Unknown partition by column: " + columnName));
+
+            return new PartitionByColumn(column.name(), false);
+          } else {
+            return new PartitionByColumn(columnAliasGenerator.uniqueAliasFor(partitionBy), true);
+          }
+        })
+        .collect(Collectors.toList());
   }
 
   private static <K> Mapper<K> buildMapper(
-      final LogicalSchema resultSchema,
-      final Optional<ColumnName> partitionByCol,
-      final PartitionByExpressionEvaluator evaluator,
+      final LogicalSchema resultSchema, // TODO: remove dead param (and simplify build())
+      final List<PartitionByColumn> partitionByCol,
+      final List<PartitionByExpressionEvaluator> evaluators,
       final ExecutionKeyFactory<K> executionKeyFactory
   ) {
-    // If partitioning by something other than an existing column, then a new key will have
-    // been synthesized. This new key must be appended to the value to make it available for
-    // stream processing, in the same way SourceBuilder appends the key and rowtime to the value:
-    final boolean appendNewKey = !partitionByCol.isPresent();
-
     return (oldK, row) -> {
-      final Object newKey = evaluator.evaluate(oldK, row);
-      final K key = executionKeyFactory.constructNewKey(oldK, genericKey(newKey));
-
-      if (row != null && appendNewKey) {
-        row.append(newKey);
+      final List<Object> newKeyComponents = evaluators.stream()
+          .map(evaluator -> evaluator.evaluate(oldK, row))
+          .collect(Collectors.toList());
+
+      final K key =
+          executionKeyFactory.constructNewKey(oldK, GenericKey.fromList(newKeyComponents));
+
+      if (row != null) {
+        for (int i = 0; i < partitionByCol.size(); i++) {
+          if (partitionByCol.get(i).shouldAppend) {
```

Comment:
instead of relying on the implicit ordering of the partition by, it might make sense to lookup the `partitionByCol.name` in the `resultSchema` at the cost of a bit of performance. thoughts?

Reply:
Do you mean replacing `shouldAppend` with a check to see whether the `partitionByCol.name` is present in the `resultSchema` as a value column, or do you mean iterating through `resultSchema` rather than `partitionByCols`? The former doesn't seem like an improvement to me, if we're still relying on the ordering of `partitionByCols` in the iteration. To remove reliance on ordering, we can do the latter and replace the lists (of columns and evaluators) with maps keyed on column name instead, but it's not clear to me that's better. It feels slightly harder to reason about code-wise but I wouldn't mind making the change. 

What's your concern regarding relying the ordering? Are you worried it's brittle, or something else?

Reply:
When I was reviewing the code I thought we could just set it at the index (e.g. `row.set(resultSchema.get(partitionCol.name).index())`). So we still keep the iteration on `partitionByCol` but we set it in the row based on it's index in the schema.

> What's your concern regarding relying the ordering? Are you worried it's brittle, or something else?

yeah - I'm worried that it's brittle.

Reply:
Ah, interesting. (We'd have to first re-size the row before calling `.set()` but that's an implementation detail.) The advantage of your proposal is that we can change the details of the result schema without needing to update the logic here. If we extended that further, really we should also be setting the existing value fields in the new row based on the result schema, rather than leaving those intact and appending key columns. 

I guess I'm not convinced this change is necessary since our test coverage in this area is quite good -- lots of tests would break if someone modified the result schema without corresponding updates here. OTOH, it's very possible I'm biased towards thinking this code is understandable as is since I've been working on it for a while. If your assessment differs as someone who hasn't worked with this code as much, I'm inclined to go with your judgment.

As for performance, LogicalSchema stores a list of columns so finding a particular column might be slow. If we implemented this we'd want to build the index mapping outside the creation of the actual mapper, and have the mapper use the index mapping directly. I'm not opposed to this. If you think it's preferable I can open a follow-up PR.